Fixed #26229 -- Improved check for model admin check admin.E124

alasdairnicol opened the pull request:
I found this tricky to get my head around, so hopefully I have got the logic correct. The existing tests pass following my change, and I have added an extra unit test.

alasdairnicol commented on tests/modeladmin/tests.py:
```diff
@@ -1553,17 +1553,33 @@ class ProductAdmin(ModelAdmin):
         class ProductAdmin(ModelAdmin):
             list_display = ['name', 'slug', 'pub_date']
-            list_editable = ['name', 'slug']
+            list_editable = ['slug', 'name']
```
Previously we were allowing `list_display[0] == list_editable[0]`. We can relax this to `list_display[0] in list_editable`. Making this change without the fix causes the old unit test to fail.

akki commented on django/contrib/admin/checks.py:
```diff
@@ -806,10 +806,10 @@ def _check_list_editable_item(self, obj, model, field_name, label):
-            # not the same. See ticket #22792 for the use case relating to this.
-            elif (obj.list_display[0] in obj.list_editable and obj.list_display[0] != obj.list_editable[0] and
-                  obj.list_display_links is not None):
+            # If list_display[0] is is list_editable, check that
```
Please remove one of the two "is".

akki replied:
"is" is occurring twice.

alasdairnicol replied:
@akki good spot, fixed now

alasdairnicol commented on django/contrib/admin/checks.py:
```diff
@@ -806,10 +806,10 @@ def _check_list_editable_item(self, obj, model, field_name, label):
+            # If list_display[0] is is list_editable, check that
+            # list_display_links is set. See ticket #22792 and #26229 for the use
+            # cases relating to this.
+            elif (obj.list_display[0] == field_name and not obj.list_display_links and obj.list_display_links is not None):
```
We call this method `_check_list_editable_item` for every field name in list_editable. Comparing `obj.list_display[0] == field_name` instead of `obj.list_display[0] in obj.list_editable` means that we can include the correct label in the error message.

The extra and not `obj.list_display_links` is required now that we don't have the `obj.list_display[0] != obj.list_editable[0]` check. If obj.list_display_links is a non-empty iterable, then we don't have to worry about the value of `list_display[0]`. There is a separate check admin.E123 to make sure that a field is not in `list_display_links` and `list_editable`.

charettes commented on django/contrib/admin/checks.py:
```diff
@@ -806,10 +806,10 @@ def _check_list_editable_item(self, obj, model, field_name, label):
+            # If list_display[0] is in list_editable, check that
+            # list_display_links is set. See tickets #22792 and #26229
+            # for use cases relating to this.
+            elif (obj.list_display[0] == field_name and not obj.list_display_links and obj.list_display_links is not None):
```
Keep line length below 119 characters.

alasdairnicol replied:
@charettes done, I've spit the elif over two lines.

alasdairnicol commented:
@gchp Please could you review this patch, since it changes the code you wrote in d8f19bb3b6.

timgraham commented on django/contrib/admin/checks.py:
```diff
@@ -806,10 +806,11 @@ def _check_list_editable_item(self, obj, model, field_name, label):
+            # list_display_links is set. See tickets #22792 and #26229
+            # for use cases relating to this.
+            elif (obj.list_display[0] == field_name and not obj.list_display_links
+                    and obj.list_display_links is not None):
```
A regression test that fails if this line is removed would be nice, otherwise it looks good to me.

timgraham commented on tests/modeladmin/tests.py:
```diff
@@ -1553,17 +1553,33 @@ class ProductAdmin(ModelAdmin):
         """
-        The first item in list_display can be the same as the first
-        in list_editable
+        The first item in list_display can be in list_editable, as
```
no comma needed and you can wrap the docstring closer to 79 characters

timgraham commented on tests/modeladmin/tests.py:
```diff
@@ -1553,17 +1553,33 @@ class ProductAdmin(ModelAdmin):
 
+    def test_list_display_first_item_in_list_editable_no_list_display_links(self):
+        """
+        The first item in list_display can not be in list_editable if
```
"cannot" is one word. add period.

alasdairnicol commented:
@timgraham I've added a regression test `test_regression_ticket_26229`, made the minor doc string fixes you suggested, and rebased the patch against master.

alasdairnicol commented:
I rebased incorrectly. Closing pull request until it's fixed.
